### PR TITLE
Add units indicator to Oracle message

### PR DESCRIPTION
**Pull request by DeliciousHair**
For price-signalled liquidation to work properly, the consumer
of the oracle message(s) must know what exchange-rate units the
oracle is using. This PR adds a filed "units" the JSON response
and in the documentation (comments in oracle.rs) indicates that
the sample oracle is using millicents. Values remain integers.

Additionally, a typo was discovered in Rocket.toml that has been
fixed up.

Resolves: #53

### CODE REVIEW

File: rocket_sqlite_oracle/Rocket.toml [hunk 1]
```diff
@@ -1,2 +1,2 @@
-[default.databases.example_signatures]
+[default.databases.signatures]
```
Comment (thomaseizinger) [inline rocket_sqlite_oracle/Rocket.toml:1]: Would have been nice to have this fix in a separate commit :)

File: rocket_sqlite_oracle/migrations/20210809053931_create_tables.sql [hunk 2]
```diff
@@ -3,6 +3,7 @@ create table if not exists signatures (
     timestamp       integer not null,
     ask_price       integer not null,
     bid_price       integer not null,
+    units           text    not null,
     ask_signature   blob    not null,
     bid_signature   blob    not null
-);
+);
```
Comment (thomaseizinger) [inline rocket_sqlite_oracle/migrations/20210809053931_create_tables.sql:9]: Missing newline at EOF. (Sublime has a setting to ensure this automatically!)

File: rocket_sqlite_oracle/src/db/kraken.rs [hunk 3]
```diff
@@ -255,9 +255,23 @@ mod wire {
...
+    pub enum Units {
+        Millicents,
+    }
+
+    impl Units {
+        pub fn units_as_str(&self) -> String {
+            match *self {
+                Units::Millicents => String::from("millicents"),
+            }
+        }
```
Comment (thomaseizinger) [inline rocket_sqlite_oracle/src/db/kraken.rs:268]: Three comments in regards to idiomatic Rust code on this snippet:

1. I would omit the `units_` part. Your struct is already called `Units` and you are taking `&self`. No need to repeat yourself here.
2. Functions that are named `as_XYZ` are meant to be cheap conversions without allocations. See https://rust-lang.github.io/api-guidelines/naming.html for reference. In this case, returning `&'static str` would be more appropriate for a function named `as_str`.
3. If you wanted a function to return `String`, a better way would be to implement `std::fmt::Display`. Every type that implements `Display`, automatically implements `ToString`. See https://doc.rust-lang.org/std/string/trait.ToString.html.

File: rocket_sqlite_oracle/src/main.rs [hunk 4]
```diff
@@ -166,6 +173,7 @@ async fn bid_price_at_time(
     let res = db::USDTickerResponse {
         timestamp: row.timestamp.try_into().unwrap(),
         price: row.price.try_into().unwrap(),
+        units: row.units,
```
Comment (thomaseizinger) [inline rocket_sqlite_oracle/src/main.rs:176]: I would have been ok with just hard-coding `millicents` as the return value here and not saving this information to the database.

File: rocket_sqlite_oracle/src/db.rs [hunk 5]
```diff
@@ -17,6 +17,7 @@ pub struct Signatures(sqlx::SqlitePool);
 pub struct USDTickerResponse {
     pub timestamp: u64,
     pub price: u64,
+    pub units: String,
```
Comment (thomaseizinger) [inline rocket_sqlite_oracle/src/db.rs:20]: In the context of this response, I think this should be singular because there is only one price.

File: rocket_sqlite_oracle/src/db/kraken.rs [hunk 6]
```diff
@@ -255,9 +255,23 @@ mod wire {
         BitcoinParseAmount(#[from] ParseAmountError),
     }
 
+    #[derive(Debug, PartialEq, Clone, Copy)]
+    pub enum Units {
+        Millicents,
+    }
```
Comment (thomaseizinger) [inline rocket_sqlite_oracle/src/db/kraken.rs:261]: Each variant of this enum only describes a single unit, hence I'd name this singular as well.

File: rocket_sqlite_oracle/migrations/20210809053931_create_tables.sql (hunk 2 above)
Comment (thomaseizinger) [reply rocket_sqlite_oracle/migrations/20210809053931_create_tables.sql:9]: Generally, it would be good to avoid diffs like this as much as possible.

File: rocket_sqlite_oracle/migrations/20210809053931_create_tables.sql [hunk 7]
```diff
@@ -3,6 +3,7 @@ create table if not exists signatures (
     timestamp       integer not null,
     ask_price       integer not null,
     bid_price       integer not null,
+    units           text    not null,
```
Comment (thomaseizinger) [inline rocket_sqlite_oracle/migrations/20210809053931_create_tables.sql:6]: Generally, editing migrations isn't really a thing that one should do.

However, in this context, I would say it is okay because this is an example so there is never really a point where we this officially released and we'd have to migrate existing databases. Instead, the migrations are only there to help people get started.

**Review by thomaseizinger — APPROVED** [review]
Great commit message! Thanks for putting the effort in :)

A few comments to mull over!

File: rocket_sqlite_oracle/src/main.rs (hunk 4 above)
Comment (DeliciousHair) [reply rocket_sqlite_oracle/src/main.rs:176]: What I wanted to do was the sensible option of:
```
        select
            timestamp,
            ask_price as price,
            "millicents" as units,
            ask_signature as signature
        from signatures
```

but that is not allowed with rocket as the migrations expect to find the `units` column in the table data. This was done this way to enable inserting in other values as required should another service with different units be employed.

File: rocket_sqlite_oracle/README.md [hunk 8]
```diff
@@ -3,7 +3,7 @@
 Do:
 
 ```bash
-$ ROCKET_DATABASES='{oracle_signatures={url="./db/signatures.sqlite"}}' cargo run -- --secret="./dev_keys/oracle_secret"
+$ cargo run -- --secret="./dev_keys/oracle_secret"
```
Comment (luckysori) [inline rocket_sqlite_oracle/README.md:6]: ❓ Is this related to the change? If not then I'd put it in a separate commit.

File: rocket_sqlite_oracle/src/db/kraken.rs [hunk 9]
```diff
@@ -257,7 +257,7 @@ mod wire {
 
     #[derive(Debug, Clone)]
     pub struct USDPrice {
-        pub millicents: u64,
+        pub exchange_rate: u64,
     }
```
Comment (luckysori) [inline rocket_sqlite_oracle/src/db/kraken.rs:261]: 🔧 Can we rename this to something like `BtcPriceInUsd` or something that conveys that we're talking about Bitcoin?

File: rocket_sqlite_oracle/src/db/kraken.rs [hunk 10]
```diff
@@ -257,7 +257,7 @@ mod wire {
 
     #[derive(Debug, Clone)]
     pub struct USDPrice {
-        pub millicents: u64,
+        pub exchange_rate: u64,
```
Comment (luckysori) [inline rocket_sqlite_oracle/src/db/kraken.rs:260]: 🙃 You might want to use a `tuple struct` in situations where you have a `struct` with a single field. As in:

```rust
struct UsdPrice(pub u64);
```

File: src/oracle.rs [hunk 11]
```diff
@@ -2,7 +2,8 @@ use elements::hashes::{sha256, Hash, HashEngine};
 use elements::secp256k1_zkp::{self, SECP256K1};
 
 pub struct Message {
-    /// Price of bitcoin in whole USD.
+    /// Price of bitcoin in millicents, where
```
Comment (luckysori) [inline src/oracle.rs:5]: 👍

**Review by luckysori — APPROVED** [review]
LGTM!
